### PR TITLE
posfilter: Add ops function to compute transducer offset from the antenna

### PR DESCRIPTION
**Pull request by madhavmk**
* Use roll pitch heading data to compute transducer offset.
* Issue #119

### CODE REVIEW

File: src/gnatss/main.py [hunk 1]
```diff
@@ -28,7 +29,7 @@
 
 
 def gather_files(
-    config: Configuration, proc: Literal["solver", "posfilter"] = "solver"
+    config: Configuration, procs: tuple = ("solver", "posfilter")
```
[the remark this answers is not in the document]
Comment (madhavmk) [reply src/gnatss/main.py:32]: Previously the `all_files_dict` returned did not contain the `roll_pitch_heading` key. Do you suggest me to call the `gather_files()` twice: once with `proc='solver'` and once with `proc=''posfilter`?

File: src/gnatss/main.py [hunk 2]
```diff
@@ -138,23 +143,45 @@ def clean_tt(
     return cleaned_travel_times
 
 
+def rotation(
```
Comment (lsetiawan) [inline src/gnatss/main.py:146]: Please move this to `ops/posfilter.py` module.

File: src/gnatss/ops/data.py [hunk 3]
```diff
@@ -40,6 +40,50 @@ def _split_cov(
     return cov
 
 
+from typing import List
```
Comment (lsetiawan) [inline src/gnatss/ops/data.py:43]: Please move this import to top of the file

File: src/gnatss/ops/data.py [hunk 4]
```diff
@@ -40,6 +40,50 @@ def _split_cov(
...
+        Columns that should be created in the df for ENU coordinates
+    array_center : ArrayCenter
+        An object containing the center of the array
+
+    Returns
+    -------
+    pd.DataFrame
+        Modified dataset with ECEF and ENU coordinates
+    """
+    import typer
```
Comment (lsetiawan) [inline src/gnatss/ops/data.py:69]: Remove this import as it doesn't appear to be used

File: src/gnatss/ops/data.py [hunk 5]
```diff
@@ -40,6 +40,50 @@ def _split_cov(
...
+    # typer.echo(f"ecef_to_enu enu_outputs: {enu_outputs}")
```
Comment (lsetiawan) [inline src/gnatss/ops/data.py:70]: Remove old debugging lines

File: src/gnatss/loaders.py [hunk 6]
```diff
@@ -170,38 +172,65 @@ def load_travel_times(
     return all_travel_times
 
 
-def load_roll_pitch_heading(files: List[str]) -> pd.DataFrame:
+def load_roll_pitch_heading(files: Union[List[str], str]) -> pd.DataFrame:
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:175]:
```suggestion
def load_roll_pitch_heading(files: "List[str] | str") -> pd.DataFrame:
```

File: src/gnatss/main.py [hunk 7]
```diff
@@ -141,20 +145,29 @@ def clean_tt(
 def get_transmit_times(
     cleaned_travel_times: pd.DataFrame,
     all_gps_solutions: pd.DataFrame,
+    rph_data: Union[pd.DataFrame, None],
```
Comment (lsetiawan) [inline src/gnatss/main.py:148]: A `Union[pd.DataFrame, None]` essentially makes this an `Optional` input. For better readability, a better practice would be `rph_data: Optional[pd.DataFrame] = None`

File: src/gnatss/main.py [hunk 8]
```diff
@@ -169,6 +182,28 @@ def get_transmit_times(
         right_on=constants.GPS_TIME,
     )
 
+    # Merge with rph data
+    if not rph_data.empty:
```
Comment (lsetiawan) [inline src/gnatss/main.py:186]: base on your typehint, doing this if wouldn't work if `rph_data` is `None` since `NoneType` doesn't have the method `.empty`. It would be better suited to check `if isinstance(rph_data, pd.DataFrame) and not rph_data.empty`.

File: src/gnatss/main.py [hunk 9]
```diff
@@ -184,24 +219,33 @@ def get_transmit_times(
 def get_reply_times(
     cleaned_travel_times: pd.DataFrame,
     all_gps_solutions: pd.DataFrame,
+    rph_data: Union[pd.DataFrame, None],
```
Comment (lsetiawan) [inline src/gnatss/main.py:222]: This is same case as `transmit_times` comments above.

File: src/gnatss/loaders.py (hunk 6 above)
Comment (madhavmk) [reply src/gnatss/loaders.py:175]: The ` | ` syntax is only supported from Python 3.10 onwards. This would fail our Python 3.9 unittests ...
References:
[3.9 Union](https://docs.python.org/tr/3.9/library/typing.html#typing.Union)
[3.10 Union](https://docs.python.org/tr/3.10/library/typing.html#typing.Union)

File: src/gnatss/main.py (hunk 7 above)
Comment (madhavmk) [reply src/gnatss/main.py:148]: I believe that `Union` makes more sense here than `Optional`, although they are equivalent. Optional gives the impression that we can choose not to input the `rph_data` argument to the `get_transmit_times` function. Whereas Union gives the impression that the `rph_data` argument can take the value as None. In our case the Optional makes sense. 
I found this FastAPI document that explains this well: [doc](https://fastapi.tiangolo.com/python-types/#using-union-or-optional)
What do you think?

File: src/gnatss/main.py (hunk 8 above)
Comment (madhavmk) [reply src/gnatss/main.py:186]: Thank you for catching this oversight 😄  . Actually the `rph_data` cannot be `None` value, and the typehint is wrong. I will update my PR with this fix

File: src/gnatss/main.py (hunk 9 above)
Comment (madhavmk) [reply src/gnatss/main.py:222]: Thank you for catching this oversight 😄 . Actually the rph_data cannot be None value, and the typehint is wrong. I will update my PR with this fix

File: src/gnatss/loaders.py [hunk 10]
```diff
@@ -170,38 +172,65 @@ def load_travel_times(
...
     """
     Loads roll pitch heading data into a pandas dataframe from a list of files.
 
     Parameters
     ----------
-    files : List[str]
-        The list of path string to files to load
+    files : Union[List[str], str]
+        The list of path string to files to load, or an empty string if no file paths found
```
[the remark this answers is not in the document]
Comment (madhavmk) [reply src/gnatss/loaders.py:182]: The default value set for the `all_files_dict` is empty string for all the keys values. If no files are found for roll_pitch_heading, quality_controls or deletions, then the key value is an empty string. 
`all_files_dict.setdefault("quality_controls", "")`
`all_files_dict.setdefault("deletions", "")`
`all_files_dict.setdefault("roll_pitch_heading", "")`

The other loaders actually also accept empty strings, so this typehint change should be applied there too.

File: src/gnatss/loaders.py (hunk 6 above)
Comment (lsetiawan) [reply src/gnatss/loaders.py:175]: Gotcha! Okay.. I forget what we support in terms of Python

File: src/gnatss/loaders.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lsetiawan) [reply src/gnatss/loaders.py:182]: Is that the behavior you're seeing by `gather_files`?

Based on this:
https://github.com/seafloor-geodesy/gnatss/blob/55ff42fea1ff4b206d443f9d1b6a935866a5c0df/src/gnatss/main.py#L53-L63

`path` is an empty string "" if it doesn't exist, see L53

However L61 ensures that it becomes a `List[str]` since we would have `all_files = [""]`

[the remark this answers is not in the document]
Comment (madhavmk) [reply src/gnatss/loaders.py:182]: Yes. That is the behavior I'm seeing in `gather_files()`. 
For illustration, I removed the `roll_pitch_heading` in the config, and printed out the value of `all_files_dict['roll_pitch_heading']`. As you can see, it is an empty string, and not a list containing an empty string ...
`
all_files_dict['roll_pitch_heading'] =
`

File: src/gnatss/main.py (hunk 7 above)
Comment (madhavmk) [reply src/gnatss/main.py]: I have modified the type to be `rph_data: pd.DataFrame`. This is because the rph_data can never be a None value. It can be an empty dataframe though!

File: src/gnatss/loaders.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lsetiawan) [reply src/gnatss/loaders.py:182]: Hmm... I'm really not seeing the behavior and that doesn't makes sense at all based on the logic of the code... if `path` variable is any string without "**" it should give a `List[str]`. I tried this locally and I shared you a screenshot of the proper behavior

I changed the path to an empty string for the test config yaml file...
```yaml
posfilter:
  atd_offsets:
    forward: 0.0053
    rightward: 0
    downward: 0.92813
  input_files:
    roll_pitch_heading:
      path: ""
```

Then using the debugger to print out the `all_files_dict` and sure enough it's getting `['']`, which is what I'm expecting


Screenshot 2024-02-13 at 2 38 07 PM


So I'm not sure what's going on, on your end to get you an empty string.. anyhow, the input to this function SHOULD NOT be an empty string, but rather a `List[str]`.

[the remark this answers is not in the document]
Comment (madhavmk) [reply src/gnatss/loaders.py:182]: As discussed in our call, I have made `load_roll_pitch_heading()` files argument accept only List[str].

File: src/gnatss/loaders.py [hunk 11]
```diff
@@ -176,32 +178,59 @@ def load_roll_pitch_heading(files: List[str]) -> pd.DataFrame:
 
     Parameters
     ----------
-    files : List[str]
-        The list of path string to files to load
+    files : Union[List[str], str]
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:181]: Please fix the docstring to reflect actual typehint

File: src/gnatss/loaders.py [hunk 12]
```diff
@@ -176,32 +178,59 @@ def load_roll_pitch_heading(files: List[str]) -> pd.DataFrame:
...
+            .reset_index(drop=True)
+            for i in files
+        ]
+        all_rph = pd.concat(rph_dfs).reset_index(drop=True)
+        return all_rph
+    else:
+        return pd.DataFrame(columns=columns)
+
+
+def load_atd_offsets(config: Configuration) -> Union[NDArray[Shape["3"], Float], None]:
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:209]:
```suggestion
def get_atd_offsets(config: Configuration) -> Union[NDArray[Shape["3"], Float], None]:
```
Since this is technically not I/O operation, let's change the function name to be "get" since we are getting the values from config

File: src/gnatss/loaders.py [hunk 13]
```diff
@@ -176,32 +178,59 @@ def load_roll_pitch_heading(files: List[str]) -> pd.DataFrame:
...
+    """
+     Loads Antenna Transducer Offset values into a numpy array.
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:211]:
```suggestion
     Retrieves the Antenna Transducer Offset values from configuration and turn them into a numpy array.
```

File: src/gnatss/main.py [hunk 14]
```diff
@@ -512,17 +580,13 @@ def load_data(all_files_dict: Dict[str, Any], config: Configuration) -> pd.DataF
         typer.echo(transponder)
     typer.echo("Finished computing harmonic mean")
 
-    # Read deletion file
-    # Set default to empty string
-    all_files_dict.setdefault("deletions", "")
+    # Read deletion file, and set default to empty list
     typer.echo("Load deletions data...")
-    cut_df = load_deletions(all_files_dict["deletions"], config=config)
+    cut_df = load_deletions(all_files_dict.get("deletions", list()), config=config)
```
Comment (lsetiawan) [inline src/gnatss/main.py:585]: Hmm... we discussed that this should be the following instead:
```python
if all_files_dict.get("deletions"):
    typer.echo("Load deletions data...")
    cut_df = load_deletions(all_files_dict["deletions"], config=config)
```

This way, then even the typer echo is not shown if deletions data is not specified

File: src/gnatss/main.py [hunk 15]
```diff
@@ -534,6 +598,12 @@ def load_data(all_files_dict: Dict[str, Any], config: Configuration) -> pd.DataF
         files=all_files_dict["travel_times"], transponder_ids=transponder_ids
     )
 
+    # Load roll-pitch-heading data, and set default to empty list
+    typer.echo("Load roll-pitch-heading data...")
+    rph_data = load_roll_pitch_heading(
+        files=all_files_dict.get("roll_pitch_heading", list())
+    )
+
```
Comment (lsetiawan) [inline src/gnatss/main.py:606]: Same thing here, please adjust to be like the deletions comments above

File: src/gnatss/main.py [hunk 16]
```diff
@@ -184,24 +219,33 @@ def get_transmit_times(
 def get_reply_times(
     cleaned_travel_times: pd.DataFrame,
     all_gps_solutions: pd.DataFrame,
+    rph_data: pd.DataFrame,
     gps_sigma_limit: float,
     transponder_ids: List[str],
+    atd_offsets: NDArray[Shape["3"], Float],
+    array_center: ArrayCenter,
```
Comment (lsetiawan) [inline src/gnatss/main.py:226]: So these two inputs are actually very specific to the computation for rph data, so I think it would make more sense to make these `Optional` rather than a required field.

**Review by lsetiawan — CHANGES_REQUESTED** [review]
Thanks for making all the changes. I think it still have a couple more minor details to be modified.

File: src/gnatss/main.py (hunk 16 above)
Comment (madhavmk) [reply src/gnatss/main.py]: Done!

File: src/gnatss/loaders.py (hunk 11 above)
Comment (lsetiawan) [inline src/gnatss/loaders.py:181]:
```suggestion
    files : List[str]
```

File: src/gnatss/loaders.py [hunk 17]
```diff
@@ -176,32 +178,59 @@ def load_roll_pitch_heading(files: List[str]) -> pd.DataFrame:
 
     Parameters
     ----------
-    files : List[str]
-        The list of path string to files to load
+    files : Union[List[str], str]
+        The list of path string to files to load, or an empty string if no file paths found
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:182]:
```suggestion
        The list of path string to files to load.
```

File: src/gnatss/main.py [hunk 18]
```diff
@@ -512,17 +585,16 @@ def load_data(all_files_dict: Dict[str, Any], config: Configuration) -> pd.DataF
...
+    if all_files_dict.get("deletions"):
+        # Read deletion file
+        typer.echo("Load deletions data...")
+        cut_df = load_deletions(all_files_dict["deletions"], config=config)
+
+    if all_files_dict.get("quality_controls"):
+        # Read quality control file
+        typer.echo("Load quality controls data...")
+        qc_df = load_quality_control(all_files_dict["quality_controls"])
+
```
Comment (lsetiawan) [inline src/gnatss/main.py:597]: revert this changes back

Comment (lsetiawan) [inline src/gnatss/main.py:597]:
```suggestion
    # Read deletion file
    # Set default to empty string
    all_files_dict.setdefault("deletions", "")
    typer.echo("Load deletions data...")
    cut_df = load_deletions(all_files_dict["deletions"], config=config)

    # Read quality control file
    # Set default to empty string
    all_files_dict.setdefault("quality_controls", "")
    typer.echo("Load quality controls data...")
    qc_df = load_quality_control(all_files_dict["quality_controls"])
```

**Review by lsetiawan — APPROVED** [review]
LGTM
